### PR TITLE
occupancy grids and creating local submaps

### PR DESCRIPTION
**Pull request by BharathMasetty**
Almost covered everything in DPG. The remaining todos are: 1) Main execution loop for DPG, 2) Visualization for grids and maps,
and 3) Joining DPG with the rest of the code.

### CODE REVIEW

**Review by mandi1267 — CHANGES_REQUESTED** [review]

File: src/dpg_slam/dpg_slam.cc [hunk 1]
```diff
@@ -602,33 +602,97 @@ namespace dpg_slam {
...
+		for (uint32_t j=0; j<(dpg_nodes_.size()-current_pass_nodes_.size()); j++) {
+			
+			DpgNode pastNode = dpg_nodes_[j];
+			std::pair<Vector2f, float> pastNodeInfo = pastNode.getEstimatedPosition();
+                	Vector2f pastNodePosition = pastNodeInfo.first;
+			float distanceToCurrentNode = (currentNodePosition-pastNodePosition).norm();
+			double overlapBetweenPastAndCurrentNode;
+			
+			// If A node from previous pass is with laser range of current node in current pass nodes, check if imporves coverage for current node.
+			if (distanceToCurrentNode < laserRange) {
```
[the remark this answers is not in the document]
Comment (BharathMasetty) [reply src/dpg_slam/dpg_slam.cc:646]: Yes. I'll make it a parameter.

File: src/dpg_slam/dpg_slam.h [hunk 2]
```diff
@@ -215,6 +252,16 @@ namespace dpg_slam {
          * Set to true if the occupancy grid should include static points.
          */
         bool include_static_;
+
+	/**
+	 * Set true of the occupancy gris is only for one node.
+	 */
+	bool is_single_node_grid_;
```
[the remark this answers is not in the document]
Comment (BharathMasetty) [reply src/dpg_slam/dpg_slam.h:259]: Not anymore. We can remove this for now.

File: src/dpg_slam/dpg_slam.cc [hunk 3]
```diff
@@ -602,33 +602,97 @@ namespace dpg_slam {
...
+    	double fractionOverlap;
+	uint32_t cellOverlapCounter = 0;
+	std::unordered_map<cellKey, CellStatus, boost::hash<cellKey>> subMapGridInfo = localSubMapGrid.getGridInfo();
+	std::unordered_map<cellKey, CellStatus, boost::hash<cellKey>> currentGridInfo = currentGrid.getGridInfo();
+	uint32_t totalCellKeys = currentGridInfo.size();
+	for(const auto &occ_grid_entry : currentGridInfo) {
+		cellKey CellToBeChecked = occ_grid_entry.first;
+		
+		// check if this key exists in localSubMapGrid
+		if (subMapGridInfo.find(CellToBeChecked) == subMapGridInfo.end()) {
```
[the remark this answers is not in the document]
Comment (BharathMasetty) [reply src/dpg_slam/dpg_slam.cc:688]: You are right. I misunderstood the logic there.

File: src/dpg_slam/dpg_slam.cc [hunk 4]
```diff
@@ -602,33 +602,97 @@ namespace dpg_slam {
...
     }
 
-    std::vector<dpgMapPoint> DpgSLAM::detectAndLabelChanges(const occupancyGrid& currGrid, const occupancyGrid& submapGrid){
-
-        std::vector<dpgMapPoint> dynamicMapPoints;
-        //TODO: Implement Alg. 1 from the paper.
-        //TODO: Bharath
-
-        return dynamicMapPoints;
+    double DpgSLAM::currentGridCoverageByLocalSubMap(const occupancyGrid &currentGrid, const occupancyGrid &localSubMapGrid){
```
[the remark this answers is not in the document]
Comment (BharathMasetty) [reply src/dpg_slam/dpg_slam.cc:678]: That's nice. I'll put that in.

File: src/dpg_slam/dpg_slam.cc [hunk 5]
```diff
@@ -602,33 +602,97 @@ namespace dpg_slam {
...
+		    						const std::vector<occupancyGrid> &poseChainGrids) {
         std::vector<DpgNode> FovNodes;
-        for (uint32_t i=0; i<(dpg_nodes_.size()-current_pass_nodes_.size()); i++) {
-            // TODO: FIll in
-        }
+
+	// Get max laser range
+	float laserRange = dpg_nodes_.back().getMeasurement().getMaxRange();
+
+	for (uint32_t i=0; i < poseChain.size(); i++) {
```
[the remark this answers is not in the document]
Comment (BharathMasetty) [reply src/dpg_slam/dpg_slam.cc:626]: Makes sense, I'll change it.

**Pull request closed without merging**